components/layout_2020/flexbox/layout.rs

```diff
@@ -2,32 +2,1090 @@
+            let total_violation: Length = unfrozen_items().map(violation).sum();
+            if total_violation == Length::zero() {
+                // “Freeze all items.”
+                // Return instead, as that’s what the next loop iteration would do.
+                let remaining_free_space =
+                    container_main_size - target_main_sizes_vec.iter().cloned().sum();
+                return (target_main_sizes_vec, remaining_free_space);
+            } else if total_violation > Length::zero() {
+                // “Freeze all the items with min violations.”
+                // “If the item’s target main size was made larger by [clamping],
+                //  it’s a min violation.”
+                for (item_and_target_main_size, frozen) in items() {
+                    if violation(item_and_target_main_size) > Length::zero() {
+                        frozen_count.set(frozen_count.get() + 1);
+                        frozen.set(true);
+                    }
+                }
+            } else {
+                // Negative total violation
+                // “Freeze all the items with max violations.”
+                // “If the item’s target main size was made smaller by [clamping],
+                //  it’s a max violation.”
+                for (item_and_target_main_size, frozen) in items() {
+                    if violation(item_and_target_main_size) < Length::zero() {
+                        frozen_count.set(frozen_count.get() + 1);
+                        frozen.set(true);
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl<'a> FlexItem<'a> {
+    // Return the hypothetical cross size together with laid out contents of the fragment.
+    // https://drafts.csswg.org/css-flexbox/#algo-cross-item
+    // “performing layout as if it were an in-flow block-level box
+    //  with the used main size and the given available space, treating `auto` as `fit-content`.”
+    fn layout(
+        &mut self,
+        used_main_size: Length,
+        flex_context: &mut FlexContext,
+    ) -> FlexItemLayoutResult {
+        match flex_context.flex_axis {
+            FlexAxis::Row => {
+                // The main axis is the container’s inline axis
+
+                // https://drafts.csswg.org/css-writing-modes/#orthogonal-flows
+                assert_eq!(
+                    flex_context.containing_block.style.writing_mode,
+                    self.box_.style().writing_mode,
+                    "Mixed writing modes are not supported yet"
+                );
+                // … and also the item’s inline axis.
+
+                match self.box_ {
+                    IndependentFormattingContext::Replaced(replaced) => {
+                        let pbm = replaced
+                            .style
+                            .padding_border_margin(flex_context.containing_block);
+                        let size = replaced.contents.used_size_as_if_inline_element(
```

Comment:
Where do we do the "treat auto as fit-content"?

Reply:
This code path is for replaced elements. I *think* `auto` and `fit-content` resolve to the same thing for those, though I might be wrong for some cases and it’s possible that we don’t implement everything in https://drafts.csswg.org/css-sizing/#intrinsic accurately.